### ISSUE TITLE
Add support for mutually recursive trait declarations

### ISSUE THREAD

**Issue by sonmarcho** [AeneasVerif/aeneas#176]
We need to add support for mutually recursive trait declarations in Aeneas. In the meantime, we should add a span to the error message.

**Comment by sonmarcho** [comment]
Another (simply recursive) example of a different flavour, inspired by the mutual recursion between `Iterator` and `IntoIterator`:
```rust
pub trait Trait {
    fn f<T : Trait>(&self, other: &T);
}
```

**Comment by oliver-butterley** [comment]
Adding a few more details to help identify when this is the underlying issue:

Currently when Aeneas encounters recursive traits it generates **"[Warn ] Mutually recursive trait declarations are not supported; the following group of mutually recursive traits is going to be extracted but their model will not type-check:..."** Often the Lean code can be made to work by manually inspecting it, understanding what is essential and finding a way to break the cyclic dependency.

**Comment by oliver-butterley** [comment]
@sonmarcho whilst recursive traits aren't supported, would it make sense that the Lean code produced by them is written to TypesExternal_Template and not to Types?

**Comment by sonmarcho** [comment]
Yes it's probably a good idea. But we would need to also extract the impls corresponding to these traits into TypesExternal_Template.

**Comment by PlanetMacro** [comment]
Fable found:

> Real-world impact datum, from a formal-verification campaign over an
> [arkworks](https://github.com/arkworks-rs/algebra)-based crypto workload
> ([PolymeshAssociation/crypto @ `zk_audit_july_2026`](https://github.com/PolymeshAssociation/crypto/tree/zk_audit_july_2026)):
> in `ark-ff` v0.5, the central
> [`Field` trait](https://github.com/arkworks-rs/algebra/blob/v0.5.0/ff/src/fields/mod.rs#L159)
> sits in a recursive declaration group with the
> [`SqrtPrecomputation<Self>` enum](https://github.com/arkworks-rs/algebra/blob/v0.5.0/ff/src/fields/sqrt.rs#L67)
> — the trait references the enum through its
> [`SQRT_PRECOMP` associated const](https://github.com/arkworks-rs/algebra/blob/v0.5.0/ff/src/fields/mod.rs#L205),
> and the enum is generic over the trait. So *every* arkworks-based
> workload loses the `Field` trait to this limitation; and once it is
> dropped, any **direct** `Field::pow`/`Field::inverse` method call in a
> translated body — e.g.
> [`sum_of_powers` in `dock_crypto_utils`](https://github.com/PolymeshAssociation/crypto/blob/zk_audit_july_2026/utils/src/ff.rs#L102-L103)
> — aborts extraction entirely with the crash reported in #1241.
> Workaround we used: rewrite `pow`/`inverse` method calls into a
> `One`-bounded recursion and a `/=` (`DivAssign`) — entry #19 of th
